File: model-client/src/main/java/org/modelix/model/lazy/LazyLoaded.java
package org.modelix.model.lazy;

public abstract class LazyLoaded<E> {

  protected String hash;
  protected IDeserializingKeyValueStore store;

  public LazyLoaded(String hash, IDeserializingKeyValueStore store) {
    this.hash = hash;
    this.store = store;
  }

  protected void init() {
    if (hash == null) {
      return;
    }
    try {
      E deserialized = store.get(hash, this::deserialize);
      init(deserialized);
    } finally {
      store = null;
      hash = null;
    }
  }

  protected abstract void init(E data);
  public abstract E deserialize(String serialized);
}

File: model-client/src/main/kotlin/org/modelix/model/lazy/LazyLoaded.kt
package org.modelix.model.lazy

import java.util.function.Function

abstract class LazyLoaded<E>(protected var hash: String?, protected var store: IDeserializingKeyValueStore?) {
    protected fun init() {
        if (hash == null) {
            return
        }
        try {
            val deserialized = store!![hash, Function { serialized: String? -> deserialize(serialized) }]
            init(deserialized)
        } finally {
            store = null
            hash = null
        }
    }

    protected abstract fun init(data: E?)
    abstract fun deserialize(serialized: String?): E
}
